rust/extns/numaflow-kafka/src/lib.rs

```diff
@@ -0,0 +1,680 @@
+                    }
+                    client_config
+                        .set("sasl.mechanisms", mechanism)
+                        .set("sasl.username", username)
+                        .set("sasl.password", password);
+                }
+            }
+        }
+
+        if let Some(tls_config) = config.tls {
+            client_config.set("security.protocol", "SSL");
+            if tls_config.insecure_skip_verify {
+                warn!(
+                    "'insecureSkipVerify' is set to true, certificate validation will not be performed when connecting to Kafka server"
+                );
+                client_config.set("enable.ssl.certificate.verification", "false");
+            }
+            if let Some(ca_cert) = tls_config.ca_cert {
+                client_config.set("ssl.ca.pem", ca_cert);
+            }
+            if let Some(client_auth) = tls_config.client_auth {
+                client_config
+                    .set("ssl.certificate.pem", client_auth.client_cert)
+                    .set("ssl.key.pem", client_auth.client_cert_private_key);
+            }
+        }
+
+        let context = KafkaContext;
+        let consumer: LoggingConsumer =
+            client_config
+                .create_with_context(context)
+                .map_err(|err| Error::Connection {
+                    server: config.brokers.join(","),
+                    error: err.to_string(),
+                })?;
+
+        // NOTE: Subscribing to a non-existent topic will not return an error
+        // The error happens only when we try to read from the topic
+        // 2025-05-09T02:45:42.239784Z ERROR rdkafka::client: librdkafka: Global error: UnknownTopicOrPartition (Broker: Unknown topic or partition): Subscribed topic not available: test-topic: Broker: Unknown topic or partition
+        // Currently, the pending returns Ok(Some(0)) if the topic is not found. When the topic is created,
+        // the consumer starts to pull messages without the need for a restart.
+        consumer
+            .subscribe(&[&config.topic])
+            .map_err(|err| Error::Kafka(format!("Failed to subscribe to topic: {}", err)))?;
+
+        // The consumer.subscribe() will not fail even if the credentials are invalid.
+        // To ensure creds/certificates are valid, we make a call to pending_messages() before starting the actor.
+        let mut actor = KafkaActor {
+            consumer,
+            read_timeout,
+            batch_size,
+            topic: config.topic,
+            handler_rx,
+        };
+
+        actor
+            .pending_messages()
+            .await
+            .map_err(|err| Error::Kafka(format!("Failed to get pending messages: {err:?}")))?;
+
+        tokio::spawn(async move {
+            tracing::info!("Starting Kafka consumer...");
+            actor.run().await;
+        });
+
+        Ok(())
+    }
+
+    async fn run(&mut self) {
```

Comment:
```suggestion
    async fn run(self) {
```

take ownership to make idiomatic?